pkg/objstore/objstore.go

```diff
@@ -194,9 +194,17 @@ func DownloadFile(ctx context.Context, logger log.Logger, bkt BucketReader, src,
 	}()
 	defer runutil.CloseWithLogOnErr(logger, f, "download block's output file")
 
-	if _, err = io.Copy(f, rc); err != nil {
+	atts, err := bkt.Attributes(ctx, src)
```

Comment:
This will double the number of API calls required to download a file. I don't think we should do it.

Reply:
if don't do it. sometimes return value of io.Copy was less than the legnth of on the object storage.
the file was not complete, but  was not err be return.
so ,the thanos compact will change state to [halt]. halt meas ....
if do it, thanos compact will retry.  
i don't know some graceful approach to fix this issue.

Reply:
> if don't do it. sometimes return value of io.Copy was less than the legnth of on the object storage.

Looks an issue with a specific backend storage that shouldn't be fixed here. Which backend storage are you using?

Reply:
S3

Reply:
I confirm I believe this is not the right place to address it, for a couple of reasons:
1. We shouldn't introduce additional API calls
2. If the reader returned by S3 `Get()` can return partial response without erroring out, then the problem is widespread across the whole codebase (any usage of `Get()`) so fixing it here would just cover a small amount of use cases

_Checking the logic to have a better proposal_

Reply:
See: https://github.com/thanos-io/thanos/issues/2805#issuecomment-778147852